[WIP] Enable raycluster autoscale

VassilisVassiliadis opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

This is a Work In Progress PR for supporting RayCluster and RayJob objects with Ray autoscaling on (current commit is just for RayCluster).

#### Which issue(s) this PR fixes:

Related to #77 but follows the methodology suggested in https://github.com/kubernetes-sigs/kueue/pull/1852#issuecomment-2675074264

#### Special notes for your reviewer:

This PR is incomplete - I've only made changes related to RayCluster and I would like some help on how to proceed. I believe the missing bit is the `kueue.x-k8s.io/standalone` label mentioned in https://github.com/kubernetes-sigs/kueue/pull/1852#issuecomment-2675074264 - I couldn't find how to use it.

When spec.enableInTreeAutoscaling is false

- The commit, maintains the current behaviour for RayCluster

When spec.enableInTreeAutoscaling is true
- The RayCluster object is associated with a Workload object that accounts only for the resources of the Head node pod
- Pods that the Ray Auto Scaler creates are annotated with the same queue as the head node
- This should have triggered Kueue to create 1 workload per worker pod - assumes "plain pod" feature flag is on.

What is missing:
- Pods that the Ray Auto Scaler creates are not currently associated with a new Workload object

For this last bit I believe that I'd have to update the logic in reconciler.go. Currently, the reconciler doesn't create a workload for the worker node pods that the ray autoscaler creates as they're not "top level jobs" and their owner is a jobtype that Kueue manages (i.e. a RayCluster).

Pods that are created for a Deployment object follow a slightly different path. I think that this is because their ownerref points to a ReplicaSet object which is not directly managed by Kueue.

I'm testing the above with the following files:

<details><summary>My `kueue-manager-config` ConfigMap</summary>

```yaml
apiVersion: v1
data:
  controller_manager_config.yaml: |
    apiVersion: config.kueue.x-k8s.io/v1beta1
    kind: Configuration
    health:
      healthProbeBindAddress: :8081
    metrics:
      bindAddress: :8443
    webhook:
      port: 9443
    leaderElection:
      leaderElect: true
      resourceName: c1f6bfd2.kueue.x-k8s.io
    controller:
      groupKindConcurrency:
        Job.batch: 5
        Pod: 5
        Workload.kueue.x-k8s.io: 5
        LocalQueue.kueue.x-k8s.io: 1
        ClusterQueue.kueue.x-k8s.io: 1
        ResourceFlavor.kueue.x-k8s.io: 1
    clientConnection:
      qps: 50
      burst: 100
    managedJobsNamespaceSelector:
      matchLabels:
        kueue-managed: "true"   # limiting Kueue to managing objects in my test namespace
    integrations:
      frameworks:
      - "batch/job"
      - "kubeflow.org/mpijob"
      - "ray.io/rayjob"
      - "ray.io/raycluster"
      - "jobset.x-k8s.io/jobset"
      - "kubeflow.org/paddlejob"
      - "kubeflow.org/pytorchjob"
      - "kubeflow.org/tfjob"
      - "kubeflow.org/xgboostjob"
      - "workload.codeflare.dev/appwrapper"
      - "pod"                            # required for RayCluster objects which autoscale
      - "deployment"
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/component: controller
    app.kubernetes.io/name: kueue
    control-plane: controller-manager
  name: kueue-manager-config
  namespace: kueue-system
```

</details>

My RayCluster:
1. enables in tree auto scaling
2. has a CPU worker group with 1 to 10 replicas
3. the expected behaviour is to see 2 workloads when creating this RayCluster (one of the head node and another for the 1 replica of the worker that gets spun up).
4. uses a localqueue called `user-queue`

<details><summary>My `RayCluster` yaml file </summary>
apiVersion: ray.io/v1
kind: RayCluster
metadata:
  labels:
    kueue.x-k8s.io/queue-name: user-queue
    controller-tools.k8s.io: "1.0"
  name: george
spec:
  enableInTreeAutoscaling: true
  rayVersion: '2.41.0'
  # Ray head pod configuration
  headGroupSpec:
    serviceType: ClusterIP
    rayStartParams:
      dashboard-host: '0.0.0.0'
    template:
      metadata:
        labels: {}
      spec:
        serviceAccountName: default
        containers:
          - name: ray-head
            image: rayproject/ray:2.41.0
            securityContext:
              runAsUser: 0
              runAsGroup: 0
            ports:
              - containerPort: 6379
                name: gcs
              - containerPort: 8265
                name: dashboard
              - containerPort: 10001
                name: client
            lifecycle:
              preStop:
                exec:
                  command: ["/bin/sh","-c","ray stop"]
            volumeMounts:
              - mountPath: /tmp/ray
                name: ray-logs
            resources:
              limits:
                cpu: "1"
                memory: "2G"
              requests:
                cpu: "1"
                memory: "2G"
        volumes:
          - name: ray-logs
            emptyDir: {}
  workerGroupSpecs:
    # the pod replicas in this group typed worker
    - replicas: 1
      minReplicas: 1
      maxReplicas: 10
      # logical group name, for this called small-group, also can be functional
      groupName: small-group
      rayStartParams:
        resources: "\"{\\\"node\\\": 1}\""
      template:
        spec:
          serviceAccountName: default
          containers:
            - name: ray-worker
              securityContext:
                runAsUser: 0
                runAsGroup: 0
              image: rayproject/ray:2.41.0
              lifecycle:
                preStop:
                  exec:
                    command: ["/bin/sh","-c","ray stop"]
              volumeMounts:
                - mountPath: /tmp/ray
                  name: ray-logs
              resources:
                limits:
                  cpu: "1"
                  memory: "1G"
                requests:
                  cpu: "1"
                  memory: "1G"
          volumes:
            - name: ray-logs
              emptyDir: {}

</details>



#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Supports for RayClusters with EnableInTreeAutoscaling on
```

mimowo commented:
This seems reasonable, but I no have double thoughts why / if we need the standalone annotation. Cause, I think the integration, when `inTreeAutoScaler` is on, would be essentially workload per pod, so as it is in Deployment, and we solve the case for Deployment just based on the queue-name label. Maybe you can check that before we conclude standalone is needed?

VassilisVassiliadis commented:
Thank you, I'll take another look at the code and write what I think the differences are between a Deployment and a RayCluster object as well as the pods they create. I'll make sure to include pointers to the code!

VassilisVassiliadis commented:
@mimowo you were right. I rebased my changes to main and added the  `kueue.x-k8s.io/managed=True` label to the pods representing worker replicas that should own a Workload object. That did the trick.

If you think my changes make sense, I can apply the same treatment to RayJob objects.

VassilisVassiliadis commented:
I'm going to take a look at the tests for the RayCluster and then apply the same logic to the RayJob. Then I'll find the docs and update them to mention the new capability. This will enable reviewers to take a look at the full picture of the PR.

mimowo commented:
/test all

VassilisVassiliadis commented:
I updated the unit tests for the RayCluster object and tidied up some poorly written comments. If this approach is good enough, I can go ahead and take a look at the docs!

VassilisVassiliadis commented:
/retest

VassilisVassiliadis commented:
@mimowo what would be the next step ?

VassilisVassiliadis commented:
Hi @PBundyra, what would be the next step for this PR ?

VassilisVassiliadis commented:
Thank you, I'll drop a ping next week! I'd like to know what's the next thing to tackle for this PR.

VassilisVassiliadis commented:
Hi @PBundyra I was wondering what would be the next step for this PR. Should I take a stab at updating the docs?

I was thinking of something along the following lines for https://kueue.sigs.k8s.io/docs/tasks/run/rayclusters/#c-limitations

---

### c. Limitations

Kueue supports autoscaling for RayCluster objects if you enable the integration of Kueue with plain pods.

- **If you opt in to Kueue managing plain Pods**, you can enable RayCluster’s internal autoscaling by setting:

  ```yaml
  spec.rayClusterSpec.enableInTreeAutoscaling: true
  ```

  In this configuration, KubeRay handles autoscaling, and Kueue creates one Workload object per worker Pod. You can define **any number** of worker groups in `spec.workerGroupSpecs`.

- **If you do not opt in to Kueue managing plain Pods**, you must disable RayCluster’s internal autoscaling by setting:

  ```yaml
  spec.rayClusterSpec.enableInTreeAutoscaling: false
  ```

  In this configuration, Kueue manages resource allocation for the entire RayCluster. You must also **limit the number of worker groups**. Since a Kueue Workload supports a maximum of 8 PodSets, your RayCluster can include **at most 7** entries in the `spec.workerGroupSpecs` field.

---

What do you think ?

VassilisVassiliadis commented:
Hi @mimowo @PBundyra is there anything I should look at to move this forward ?

PBundyra commented:
Hi @VassilisVassiliadis, thanks for working on that and sorry for the late reply. Since this is important step for Kueue I think the best people to do a first pass are @mimowo and @tenzen-y. Mimowo is currently out of office and will be back in September

VassilisVassiliadis commented:
Thank you Patryk!

VassilisVassiliadis commented:
This sounds interesting.

Give me a few days to look into how KubeRay handles autoscaling because I can see a couple of ways this could work. I think it all depends on how the ray autoscaler handles k8s rejecting its patches to the RayCluster object.

VassilisVassiliadis commented:
I started working on this here: https://github.com/VassilisVassiliadis/kueue/tree/77_elasticjob_for_raycluster

The good news is that it seems to be working fine. I can see new pods when I scale up the RayCluster, first they get schedule-gated and then when there's enough room in the ClusterQueue they transition to pending/running. When I scale down the workers in the RayCluster I see pods getting deleted and others staying where they are. In both cases the Workload objects get updated.

The weird news is that I didn't have to update the `SetupIndexes()` function of the RayCluster. Based on this comment https://github.com/ichekrygin/kueue/blob/f1d44095fcd85f0c1c394c6a32faf5cdbeccab30/pkg/controller/jobs/job/job_controller.go#L373-L379 (from the PR https://github.com/kubernetes-sigs/kueue/pull/5510) this is something that's required. This makes me think that I'm missing something but I'm running out of time to dig into this deeper today.

So, I'm going to add a couple of unit tests next week and do a bit more testing locally before opening a PR. For that, I'll have to take a closer look at the relevant unit-tests in batchv1/Job for WorkloadSlicing as I haven't tested functionality like that before.

@tenzen-y when the time comes to open my PR, should I ask you to you review it ?

VassilisVassiliadis commented:
Closing this because it's superseded by https://github.com/kubernetes-sigs/kueue/pull/6662
